mempool/src/pool/reorg.rs

```diff
@@ -0,0 +1,162 @@
+    #[error("Could not find the previous tip")]
+    OldTip,
+    #[error("Could not find the previous tip index")]
+    OldTipIndex,
+    #[error("Could not find the new tip index")]
+    NewTipIndex,
+    #[error("Block {0:?} not found while traversing history")]
+    BlockNotFound(Id<Block>),
+    #[error("Chainstate call: {0}")]
+    ChainstateCall(#[from] subsystem::subsystem::CallError),
+}
+
+/// Iterate over blocks following the parent link. Apply given function to each block.
+fn for_each_block(
+    chainstate: &dyn ChainstateInterface,
+    mut curr_id: Id<Block>,
+    stop_id: Id<GenBlock>,
+    mut func: impl FnMut(Block),
+) -> Result<(), ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    while curr_id != stop_id {
+        let block = chainstate
+            .get_block(curr_id)?
+            .ok_or_else(|| ReorgError::BlockNotFound(curr_id))?;
+        curr_id = match block.prev_block_id().classify(chain_config) {
+            common::chain::GenBlockId::Genesis(_) => break,
+            common::chain::GenBlockId::Block(block_id) => block_id,
+        };
+        func(block);
+    }
+    Ok(())
+}
+
+fn find_disconnected_txs(
+    chainstate: &dyn ChainstateInterface,
+    old_tip_id: Id<GenBlock>,
+    new_tip_id: Id<Block>,
+) -> Result<Vec<SignedTransaction>, ReorgError> {
+    let chain_config = chainstate.get_chain_config();
+    let old_tip_id = match old_tip_id.classify(chain_config) {
+        common::chain::GenBlockId::Genesis(_) => return Ok(Vec::new()),
+        common::chain::GenBlockId::Block(id) => id,
+    };
+
+    let common_id = {
+        let old_index = chainstate.get_block_index(&old_tip_id)?.ok_or(ReorgError::OldTipIndex)?;
+        let new_index = chainstate.get_block_index(&new_tip_id)?.ok_or(ReorgError::NewTipIndex)?;
+        let common_index = chainstate.last_common_ancestor(&old_index.into(), &new_index.into())?;
+        common_index.block_id()
+    };
+
+    // Short circuit the expensive processing below if no blocks are being disconnected
+    if old_tip_id == common_id {
+        return Ok(Vec::new());
+    }
+
+    // Collect IDs of trancsactions included in the newly connected chain
+    let mut connected_txs = BTreeSet::new();
+    for_each_block(chainstate, new_tip_id, common_id, |block| {
+        connected_txs.extend(block.transactions().iter().map(|t| t.transaction().get_id()))
+    })?;
+
+    // Collect txns that have been removed from the blockchain and not added on the new fork
+    let mut disconnected_txs = Vec::new();
+    for_each_block(chainstate, old_tip_id, common_id, |block| {
+        // We iterate blocks in the reverse order, so for consistent transaction input/output
+        // dependencies, take transacitons in reverse too,
+        let txns = block.transactions().iter().rev();
+        // We disregerd transactions that are re-added on the newly connected chain
+        let txns = txns.filter(|txn| !connected_txs.contains(&txn.transaction().get_id()));
```

Comment:
I'm not too sure why this filter is being done. The presence of a transaction in two branches should be an issue. It's completely valid. In fact, this is very common. When two miners/stakers compete to create a block, usually both of them same the same mempool and share a large fraction of the available transactions. Meaning: A transaction being disconnected on one branch and reconnected in the new branch should be OK.

Reply:
I guess what's meant here is that the same transaction is already in chainstate... maybe this is OK. 

Geez this process is too complex... it's supposed to be simple and composable... but I see you're trying to do an optimization by not doing it "block by block"...

Reply:
One thing I've been thinking about... to be clear this doesn't necessarily mean that the transactions you're adding to the mempool will be accepted. This is because there could be a conflict. A conflict is defined as two different transactions spending same output (basically what double-spend does). If every branch has a different transaction spending a different output, then adding transactions will not work. Having this in mind, we gotta ask whether this complicated processing is necessary... who cares? We could just add the transactions from the disconnected branch to the transaction verifier, and if the transaction already exists in chainstate (by merit of a double-spend), then it will be rejected anyway. How much value does it add to attempt to collect ids, filter, etc? My bet is none.

Reply:
Yea this is basically meant to be an optimisation. It avoids checking transactions that would have been rejected anyway. The idea is that collecting IDs is proportional to the depth of the reorg while looking up stuff in the utxo set is probably rather expensive as it can grow very large over time (proportional to the entire history worst case). In addition to that, the "hit rate" of this optimization is likely to be very high.

It's hard to know how it works out in the end without ptoper benchmarking. I can imagine the difference starts being significant only once the UTXO set reaches certain critical size. I think bitcoin does something like this.

Also I don't think it's particularly complicated, just collecting some IDs and filtering them out...

Reply:
One thing that could be done to make it easier to read, is create function that does the collection and another function that does the filtering, and add documentation to them explaining the story.

Reply:
Attempted to streamline this a bit. Let me know if it looks better.

Reply:
Much nicer now! Thank you!